compiler/rustc_expand/src/expand.rs

```diff
@@ -2315,7 +2315,17 @@ impl<'a, 'b> MutVisitor for InvocationCollector<'a, 'b> {
     }
 
     fn visit_ty(&mut self, node: &mut P<ast::Ty>) {
-        self.visit_node(node)
+        // Save the pre-expanded name of this `ImplTrait`, so that later when defining
+        // an APIT we use a name that doesn't have any placeholder fragments in it.
+        if let ast::TyKind::ImplTrait(id, _) = &mut node.kind {
+            // HACK: Assign an ID to this node out-of-order.
+            self.visit_id(id);
+            let id = *id;
+            let name = Symbol::intern(&pprust::ty_to_string(node).replace('\n', " "));
+            self.cx.resolver.insert_impl_trait_name(id, name);
+        }
```

Comment:
I dislike this part in particular.

Perhaps the `insert_impl_trait_name` call can be moved to `InvocationCollectorNode::walk` on types?
In that case the node ID will already be assigned.
It will be also be a more correct location to do this in general (e.g. compatible with supporting `#[cfg]` and macro attributes on types).

Reply:
If I move it to `InvocationCollectorNode::walk`, then all I have access to is a `MutVisitor`, so I can't actually call something like `insert_impl_trait_name`.